### PR TITLE
rook-ceph: Allow user to provide sub-component resources

### LINKED ISSUE #1476: Rook: Ceph OSDs consume up to 12 GB memory
tl;dr: Provide a way for the user where they can limit the memory and CPU of the ceph sub-components like OSD, MGR, MON, etc.

---

Right now there is no way for a user to specify memory limits on the OSD or any other sub-components of Rook. Since there are no resource limits specified the pod uses all the memory availble on the host.

As you can see that the following OSD deployment has no Kubernetes resources (memory or cpu and limits or request) set. But still the env vars are trying to reference them. So here empty values are being referenced:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
  name: rook-ceph-osd-28
  namespace: rook
...
      containers:
      - args:
        - --foreground
        - --id
        - "28"
        - --fsid
        - 77d48d93-351e-4a00-a554-d99e625914d7
        - --setuser
        - ceph
        - --setgroup
        - ceph
        - --crush-location=root=default host=lokomotive-production-storage-worker-2
          region=ewr1
        - --log-to-stderr=true
        - --err-to-stderr=true
        - --mon-cluster-log-to-stderr=true
        - '--log-stderr-prefix=debug '
        - --default-log-to-file=false
        - --default-mon-cluster-log-to-file=false
        - --ms-learn-addr-from-peer=false
        command:
        - ceph-osd
        env:
        - name: POD_MEMORY_LIMIT
          valueFrom:
            resourceFieldRef:
              divisor: "0"
              resource: limits.memory
        - name: POD_MEMORY_REQUEST
          valueFrom:
            resourceFieldRef:
              divisor: "0"
              resource: requests.memory
        - name: POD_CPU_LIMIT
          valueFrom:
            resourceFieldRef:
              divisor: "1"
              resource: limits.cpu
        - name: POD_CPU_REQUEST
          valueFrom:
            resourceFieldRef:
              divisor: "0"
              resource: requests.cpu
        image: ceph/ceph:v15.2.5-20200916
        name: osd
        resources: {}
```

One would expect that empty values are populated inside the pod, but if you see the env vars all the limit values are capped at the host limits:

```bash
POD_MEMORY_LIMIT=135039045632
POD_MEMORY_REQUEST=0
POD_CPU_LIMIT=32
POD_CPU_REQUEST=0
```

This host has 125 GB memory and 32 cores of CPU.

Above automatically reflects in the OSD config:

```
[root@rook-ceph-tools-6f7fccd4d6-fvgbk /]# ceph tell osd.28 config show | grep memory
    "osd_memory_base": "805306368",
    "osd_memory_cache_min": "134217728",
    "osd_memory_cache_resize_interval": "1.000000",
    "osd_memory_expected_fragmentation": "0.150000",
    "osd_memory_target": "108031236505",
    "osd_memory_target_cgroup_limit_ratio": "0.800000",
```

`osd_memory_target` is the amount of memory that OSD is allowed to expand upto. And it is appropriately set proportional to the host memory limit.

```
osd_memory_target = floor( POD_MEMORY_LIMIT * osd_memory_target_cgroup_limit_ratio )
```

### PR DESCRIPTION
**Pull request by surajssd**
This PR adds a new parameter `resources` for sub-components' cpu and
memory requests and limits.

Fixes #1476

### CODE REVIEW

File: assets/terraform-modules/tinkerbell-sandbox/assets/deploy/tls/Dockerfile [hunk 1]
```diff
@@ -1,7 +1,8 @@
-FROM alpine:3.11
+FROM registry.fedoraproject.org/fedora:34
```
Comment (ipochi) [inline assets/terraform-modules/tinkerbell-sandbox/assets/deploy/tls/Dockerfile:1]: @surajssd It doesn't belong in this PR, would it be trouble to take it out in a separate PR. ?

Comment (surajssd) [reply assets/terraform-modules/tinkerbell-sandbox/assets/deploy/tls/Dockerfile:1]: Sure, doing it.

File: pkg/components/internal/testutil/jsonpath.go [hunk 2]
```diff
@@ -103,6 +104,29 @@ func MatchJSONPathInt64Value(t *testing.T, yamlConfig string, jsonPath string, e
 	}
 }
 
+// MatchJSONPathJSONValue is a helper function for component unit tests. It compares the JSON values
```
Comment (ipochi) [inline pkg/components/internal/testutil/jsonpath.go:107]: Any new util functions that need not be visible should be part of the internal package 

example this file : `internal/util_test.go`

**Review by ipochi — COMMENTED** [review]
Are there reasonable examples out on in the world that provide resource limits/requests for the Rook osd, monitor, prepare etc jobs ? I am merely asking whats the need for this feature ? Has there been instances where the resources in question are consuming lots of cpu and memory ?

If so, might we also suggest sane values to use in the configuration reference document or the  the existing how to guide for configuring rook/rook-ceph (

**Comment by surajssd** [comment]
> Are there reasonable examples out on in the world that provide resource limits/requests for the Rook osd, monitor, prepare etc jobs ? I am merely asking whats the need for this feature ? Has there been instances where the resources in question are consuming lots of cpu and memory ?

Yes, the linked issue describes the problem: https://github.com/kinvolk/lokomotive/issues/1476.

> If so, might we also suggest sane values to use in the configuration reference document or the the existing how to guide for configuring rook/rook-ceph

Yes,this is a good idea to point folks to https://rook.io/docs/rook/v1.6/ceph-cluster-crd.html#cluster-wide-resources-configuration-settings. But instead of doing it via how-to-guide, I think pointing folks from the config ref makes it more accessible. How-to-guide by the nature of it could hide the details.

File: pkg/components/internal/testutil/jsonpath.go (hunk 2 above)
Comment (surajssd) [reply pkg/components/internal/testutil/jsonpath.go:107]: But this `pkg/components/internal/testutil/jsonpath.go` has all the relevant code that deals with finding values at JSON path, converting into appropriate concrete data types and comparing values.

Comment (surajssd) [reply pkg/components/internal/testutil/jsonpath.go:107]: And it will be needed in various `_test.go` files, I don't fully understand what are you suggesting?

Comment (ipochi) [reply pkg/components/internal/testutil/jsonpath.go:107]: I'm suggesting to use the internal package for utility functions such as the one below

https://github.com/kinvolk/lokomotive/blob/master/internal/util_test.go

**Comment by surajssd** [comment]
@ipochi I have addressed your concerns regarding documentation, PTAL.

**Review by ipochi — APPROVED** [review]
LGTM
